File: 20150823-StringCalculator-Scala/src/main/scala/kata/SC.scala
package kata

import java.lang.Integer.parseInt
import java.util.regex.Pattern.quote

object SC {
	def add(input: String): Int = {
		val nums = parse(input)

		if (nums.exists(_ < 0))
			throw new IllegalArgumentException(nums.filter(_ < 0).mkString(","))
		else
			nums.filter(_ < 1001).sum
	}

	private def parse(input: String): Array[Int] = tokenize(input).map(parseInt)

	private def tokenize(input: String): Array[String] = {
		val SingleCharacterSeparator = "//(.)\n(.+)".r
		val MultiCharacterSeparator = "//\\[(.+)\\]\n(.+)".r

		input match {
			case "" => Array[String]()
			case SingleCharacterSeparator(sep, nums) => nums.split(quote(sep))
			case MultiCharacterSeparator(seps, nums) => nums.split(seps.split("\\]\\[").map(quote).mkString("|"))
			case _ => input.split(",|\n")
		}
	}
}

File: 20160115-StringCalculator-Java8/src/main/java/kata/SC.java
package kata;

import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.regex.Pattern.quote;

public class SC {
    public static int add(String input) {
        if (parse(input).anyMatch(n -> n < 0)) {
            throw new IllegalArgumentException(parse(input).filter(n -> n < 0).map(Object::toString).collect(Collectors.joining(",")));
        } else {
            return parse(input).filter(n -> n < 1001).reduce(0, (accumulator, n) -> accumulator + n);
        }
    }

    private static Stream<Integer> parse(String input) {
        if (input.isEmpty()) {
            return tokenize("0", ",");
        } else if (input.startsWith("//[")) {
            final int indexOfNewline = input.indexOf('\n');
            final String separatorsRegex = Stream.of(input.substring(3, indexOfNewline - 1).split("\\]\\[")).sorted((x, y) -> y.length() - x.length()).map(Pattern::quote).collect(Collectors.joining("|"));

            return tokenize(input.substring(indexOfNewline + 1), separatorsRegex);
        } else if (input.startsWith("//")) {
            return tokenize(input.substring(4), quote(input.substring(2, 3)));
        } else {
            return tokenize(input, ",|\n");
        }
    }

    private static Stream<Integer> tokenize(String numbers, String separatorsRegex) {
        return Stream.of(numbers.split(separatorsRegex)).map(Integer::parseInt);
    }
}
